stress: failed test in cockroach/rpc/rpc.test: TestCircuitBreaker

Binary: cockroach/static-tests.tar.gz sha: https://github.com/cockroachdb/cockroach/commits/83ca90e90d581fd99a28ea20be7d3a47870ef378

Stress build found a failed test:

```
=== RUN   TestCircuitBreaker
E160806 04:49:19.729653 rpc/context.go:191  rpc error: code = 12 desc = unknown service cockroach.rpc.Heartbeat
E160806 04:49:19.776859 rpc/context.go:191  rpc error: code = 12 desc = unknown service cockroach.rpc.Heartbeat
E160806 04:49:19.858298 rpc/context.go:191  rpc error: code = 12 desc = unknown service cockroach.rpc.Heartbeat
E160806 04:49:19.979764 rpc/context.go:191  rpc error: code = 12 desc = unknown service cockroach.rpc.Heartbeat
E160806 04:49:20.246936 rpc/context.go:191  rpc error: code = 12 desc = unknown service cockroach.rpc.Heartbeat
E160806 04:49:20.787734 rpc/context.go:191  rpc error: code = 12 desc = unknown service cockroach.rpc.Heartbeat
E160806 04:49:21.783398 rpc/context.go:191  rpc error: code = 12 desc = unknown service cockroach.rpc.Heartbeat
E160806 04:49:22.800223 rpc/context.go:191  rpc error: code = 12 desc = unknown service cockroach.rpc.Heartbeat
E160806 04:49:23.785727 rpc/context.go:191  rpc error: code = 12 desc = unknown service cockroach.rpc.Heartbeat
E160806 04:49:24.788369 rpc/context.go:191  rpc error: code = 12 desc = unknown service cockroach.rpc.Heartbeat
E160806 04:49:25.790459 rpc/context.go:191  rpc error: code = 12 desc = unknown service cockroach.rpc.Heartbeat
E160806 04:49:26.787557 rpc/context.go:191  rpc error: code = 12 desc = unknown service cockroach.rpc.Heartbeat
E160806 04:49:27.788387 rpc/context.go:191  rpc error: code = 12 desc = unknown service cockroach.rpc.Heartbeat
E160806 04:49:28.790563 rpc/context.go:191  rpc error: code = 12 desc = unknown service cockroach.rpc.Heartbeat
E160806 04:49:29.789964 rpc/context.go:191  rpc error: code = 12 desc = unknown service cockroach.rpc.Heartbeat
E160806 04:49:30.785989 rpc/context.go:191  rpc error: code = 12 desc = unknown service cockroach.rpc.Heartbeat
E160806 04:49:31.801600 rpc/context.go:191  rpc error: code = 12 desc = unknown service cockroach.rpc.Heartbeat
E160806 04:49:32.793040 rpc/context.go:191  rpc error: code = 12 desc = unknown service cockroach.rpc.Heartbeat
E160806 04:49:33.793717 rpc/context.go:191  rpc error: code = 12 desc = unknown service cockroach.rpc.Heartbeat
--- FAIL: TestCircuitBreaker (15.11s)
	testing.go:117: rpc/context_test.go:419, condition failed to evaluate within 15s: breaker not open: <nil>
```

Run Details:

```
117 runs so far, 0 failures, over 5s
240 runs so far, 0 failures, over 10s
362 runs so far, 0 failures, over 15s
487 runs so far, 0 failures, over 20s
609 runs so far, 0 failures, over 25s
731 runs so far, 0 failures, over 30s
856 runs so far, 0 failures, over 35s
981 runs so far, 0 failures, over 40s
1102 runs so far, 0 failures, over 45s
1223 runs so far, 0 failures, over 50s
1346 runs so far, 0 failures, over 55s
1467 runs so far, 0 failures, over 1m0s
1594 runs so far, 0 failures, over 1m5s
1718 runs so far, 0 failures, over 1m10s
1840 runs so far, 0 failures, over 1m15s
1961 runs so far, 0 failures, over 1m20s
2083 runs so far, 0 failures, over 1m25s
2209 runs so far, 0 failures, over 1m30s
2331 runs so far, 0 failures, over 1m35s
2453 runs so far, 0 failures, over 1m40s
2579 runs so far, 0 failures, over 1m45s
2701 runs so far, 0 failures, over 1m50s
2824 runs so far, 0 failures, over 1m55s
2943 runs so far, 0 failures, over 2m0s
3070 runs so far, 0 failures, over 2m5s
3190 runs so far, 0 failures, over 2m10s
3315 runs so far, 0 failures, over 2m15s
3435 runs so far, 0 failures, over 2m20s
3562 runs so far, 0 failures, over 2m25s
3681 runs so far, 0 failures, over 2m30s
3804 runs so far, 0 failures, over 2m35s
3930 runs so far, 0 failures, over 2m40s
4051 runs so far, 0 failures, over 2m45s
4175 runs so far, 0 failures, over 2m50s
4300 runs so far, 0 failures, over 2m55s
4420 runs so far, 0 failures, over 3m0s
4545 runs so far, 0 failures, over 3m5s
4668 runs so far, 0 failures, over 3m10s
4791 runs so far, 0 failures, over 3m15s
4915 runs so far, 0 failures, over 3m20s
5034 runs so far, 0 failures, over 3m25s
5162 runs so far, 0 failures, over 3m30s
5281 runs so far, 0 failures, over 3m35s
5404 runs so far, 0 failures, over 3m40s
5527 runs so far, 0 failures, over 3m45s
5652 runs so far, 0 failures, over 3m50s
5770 runs so far, 0 failures, over 3m55s
5897 runs so far, 0 failures, over 4m0s
6021 runs so far, 0 failures, over 4m5s
6142 runs so far, 0 failures, over 4m10s
6265 runs so far, 0 failures, over 4m15s
6389 runs so far, 0 failures, over 4m20s
6510 runs so far, 0 failures, over 4m25s
6630 runs so far, 0 failures, over 4m30s
6755 runs so far, 0 failures, over 4m35s
6880 runs so far, 0 failures, over 4m40s
7000 runs so far, 0 failures, over 4m45s
7121 runs so far, 0 failures, over 4m50s
7244 runs so far, 0 failures, over 4m55s
7365 runs so far, 0 failures, over 5m0s
7491 runs so far, 0 failures, over 5m5s
7615 runs so far, 0 failures, over 5m10s
7735 runs so far, 0 failures, over 5m15s
7856 runs so far, 0 failures, over 5m20s
7980 runs so far, 0 failures, over 5m25s
8101 runs so far, 0 failures, over 5m30s
8221 runs so far, 0 failures, over 5m35s
8343 runs so far, 0 failures, over 5m40s
8410 runs completed, 1 failures, over 5m43s
FAIL
```
Please assign, take a look and update the issue accordingly.

--- a/rpc/breaker.go
+++ b/rpc/breaker.go
@@ -20,12 +20,52 @@ import (
 	"time"
 
 	"github.com/cenk/backoff"
+	"github.com/cockroachdb/cockroach/util/hlc"
+	"github.com/facebookgo/clock"
 	circuit "github.com/rubyist/circuitbreaker"
 )
 
+// breakerClock is an implementation of clock.Clock that internally uses an
+// hlc.Clock. It is used to bridge the hlc clock to the circuit breaker
+// clocks. Note that it only implements the After() and Now() methods needed by
+// circuit breakers and backoffs.
+type breakerClock struct {
+	clock *hlc.Clock
+}
+
+func (c *breakerClock) After(d time.Duration) <-chan time.Time {
+	return time.After(d)
+}
+
+func (c *breakerClock) AfterFunc(d time.Duration, f func()) *clock.Timer {
+	panic("unimplemented")
+}
+
+func (c *breakerClock) Now() time.Time {
+	return c.clock.PhysicalTime()
+}
+
+func (c *breakerClock) Sleep(d time.Duration) {
+	panic("unimplemented")
+}
+
+func (c *breakerClock) Tick(d time.Duration) <-chan time.Time {
+	panic("unimplemented")
+}
+
+func (c *breakerClock) Ticker(d time.Duration) *clock.Ticker {
+	panic("unimplemented")
+}
+
+func (c *breakerClock) Timer(d time.Duration) *clock.Timer {
+	panic("unimplemented")
+}
+
+var _ clock.Clock = &breakerClock{}
+
 // newBackOff creates a new exponential backoff properly configured for RPC
 // connection backoff.
-func newBackOff() backoff.BackOff {
+func newBackOff(clock backoff.Clock) backoff.BackOff {
 	// This exponential backoff limits the circuit breaker to 1 second
 	// intervals between successive attempts to resolve a node address
 	// and connect via GRPC.
@@ -43,17 +83,16 @@ func newBackOff() backoff.BackOff {
 		Multiplier:          1.5,
 		MaxInterval:         1 * time.Second,
 		MaxElapsedTime:      0,
-		Clock:               backoff.SystemClock,
+		Clock:               clock,
 	}
 	b.Reset()
 	return b
 }
 
-// NewBreaker creates a new circuit breaker properly configured for RPC
-// connections.
-func NewBreaker() *circuit.Breaker {
+func newBreaker(clock clock.Clock) *circuit.Breaker {
 	return circuit.NewBreakerWithOptions(&circuit.Options{
-		BackOff:    newBackOff(),
+		BackOff:    newBackOff(clock),
+		Clock:      clock,
 		ShouldTrip: circuit.ThresholdTripFunc(1),
 	})
 }

--- a/rpc/context.go
+++ b/rpc/context.go
@@ -78,6 +78,7 @@ type Context struct {
 	*base.Context
 
 	localClock   *hlc.Clock
+	breakerClock breakerClock
 	Stopper      *stop.Stopper
 	RemoteClocks *RemoteClockMonitor
 
@@ -95,17 +96,20 @@ type Context struct {
 }
 
 // NewContext creates an rpc Context with the supplied values.
-func NewContext(baseCtx *base.Context, clock *hlc.Clock, stopper *stop.Stopper) *Context {
+func NewContext(baseCtx *base.Context, hlcClock *hlc.Clock, stopper *stop.Stopper) *Context {
 	ctx := &Context{
 		Context: baseCtx,
 	}
-	if clock != nil {
-		ctx.localClock = clock
+	if hlcClock != nil {
+		ctx.localClock = hlcClock
 	} else {
 		ctx.localClock = hlc.NewClock(hlc.UnixNano)
 	}
+	ctx.breakerClock = breakerClock{
+		clock: ctx.localClock,
+	}
 	ctx.Stopper = stopper
-	ctx.RemoteClocks = newRemoteClockMonitor(clock, 10*defaultHeartbeatInterval)
+	ctx.RemoteClocks = newRemoteClockMonitor(ctx.localClock, 10*defaultHeartbeatInterval)
 	ctx.HeartbeatInterval = defaultHeartbeatInterval
 	ctx.HeartbeatTimeout = 2 * defaultHeartbeatInterval
 	ctx.conns.cache = make(map[string]connMeta)
@@ -170,7 +174,7 @@ func (ctx *Context) GRPCDial(target string, opts ...grpc.DialOption) (*grpc.Clie
 
 	breaker, ok := ctx.conns.breakers[target]
 	if !ok {
-		breaker = NewBreaker()
+		breaker = ctx.NewBreaker()
 		ctx.conns.breakers[target] = breaker
 	}
 	if !breaker.Ready() {
@@ -214,6 +218,12 @@ func (ctx *Context) GRPCDialOption() (grpc.DialOption, error) {
 		dialOpt = grpc.WithTransportCredentials(credentials.NewTLS(tlsConfig))
 	}
 	return dialOpt, nil
+}
+
+// NewBreaker creates a new circuit breaker properly configured for RPC
+// connections.
+func (ctx *Context) NewBreaker() *circuit.Breaker {
+	return newBreaker(&ctx.breakerClock)
 }
 
 // setConnHealthy sets the health status of the connection.

--- a/rpc/context_test.go
+++ b/rpc/context_test.go
@@ -26,10 +26,13 @@ import (
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/credentials"
 
+	"golang.org/x/net/context"
+
 	"github.com/cockroachdb/cockroach/testutils"
 	"github.com/cockroachdb/cockroach/util"
 	"github.com/cockroachdb/cockroach/util/hlc"
 	"github.com/cockroachdb/cockroach/util/leaktest"
+	"github.com/cockroachdb/cockroach/util/log"
 	"github.com/cockroachdb/cockroach/util/netutil"
 	"github.com/cockroachdb/cockroach/util/stop"
 	"github.com/cockroachdb/cockroach/util/syncutil"
@@ -397,7 +400,9 @@ func TestCircuitBreaker(t *testing.T) {
 	stopper := stop.NewStopper()
 	defer stopper.Stop()
 
-	clock := hlc.NewClock(time.Unix(0, 1).UnixNano)
+	// We use a manual clock that we never increment so the circuit breakers do
+	// not move from open to half-open due to timeout.
+	clock := hlc.NewClock(hlc.NewManualClock(1).UnixNano)
 	serverCtx := newNodeTestContext(clock, stopper)
 	_, ln := newTestServer(t, serverCtx, true)
 	remoteAddr := ln.Addr().String()
@@ -408,6 +413,35 @@ func TestCircuitBreaker(t *testing.T) {
 		t.Fatal(err)
 	}
 
+	clientCtx.conns.Lock()
+	events := make(chan circuit.ListenerEvent, 10)
+	clientCtx.conns.breakers[remoteAddr].AddListener(events)
+	clientCtx.conns.Unlock()
+
+	ctx := context.Background()
+	done := make(chan struct{})
+	go func() {
+		for {
+			select {
+			case e := <-events:
+				switch e.Event {
+				case circuit.BreakerTripped:
+					log.Infof(ctx, "BreakerTripped")
+				case circuit.BreakerReset:
+					log.Infof(ctx, "BreakerReset")
+				case circuit.BreakerFail:
+					log.Infof(ctx, "BreakerFail")
+				case circuit.BreakerReady:
+					log.Infof(ctx, "BreakerReady")
+				default:
+					log.Infof(ctx, "%d", e.Event)
+				}
+			case <-done:
+				return
+			}
+		}
+	}()
+
 	// Wait until the breaker opens. This will occur when the heartbeat fails
 	// (because there is no heartbeat service).
 	util.SucceedsSoon(t, func() error {
@@ -415,11 +449,21 @@ func TestCircuitBreaker(t *testing.T) {
 		if err == circuit.ErrBreakerOpen {
 			return nil
 		}
-		return errors.Errorf("breaker not open: %v", err)
+		clientCtx.conns.Lock()
+		breaker := clientCtx.conns.breakers[remoteAddr]
+		err = errors.Errorf("breaker not open: %v: ready=%t tripped=%t failures=%d/%d successes=%d",
+			err, breaker.Ready(), breaker.Tripped(), breaker.Failures(),
+			breaker.ConsecFailures(), breaker.Successes())
+		log.Infof(ctx, "%s", err)
+		clientCtx.conns.Unlock()
+		return err
 	})
+
+	close(done)
 
 	// Reset the breaker. The next dial will succeed.
 	clientCtx.conns.Lock()
+	log.Infof(ctx, "resetting breaker")
 	clientCtx.conns.breakers[remoteAddr].Reset()
 	clientCtx.conns.Unlock()
 	if _, err := clientCtx.GRPCDial(remoteAddr); err != nil {

--- a/storage/raft_transport.go
+++ b/storage/raft_transport.go
@@ -192,7 +192,7 @@ func (t *RaftTransport) getNodeConn(nodeID roachpb.NodeID) *grpc.ClientConn {
 	t.mu.Lock()
 	breaker, ok := t.mu.breakers[nodeID]
 	if !ok {
-		breaker = rpc.NewBreaker()
+		breaker = t.rpcContext.NewBreaker()
 		t.mu.breakers[nodeID] = breaker
 	}
 	t.mu.Unlock()

